src/storage/mvcc/lock.rs

```diff
@@ -54,15 +57,54 @@ impl LockType {
     }
 }
 
+#[derive(new, PartialEq, Copy, Clone, Debug)]
+pub enum TransactionKind {
+    Optimistic,
+    Pessimistic(NonZeroTimeStamp),
+}
+
+impl TransactionKind {
+    fn write_bytes(self, b: &mut Vec<u8>) {
+        if let TransactionKind::Pessimistic(ts) = self {
+            b.push(FOR_UPDATE_TS_PREFIX);
+            b.encode_u64(ts.into_inner()).unwrap();
+        }
+    }
+
+    pub fn is_pessimistic(self) -> bool {
+        match self {
+            TransactionKind::Optimistic => false,
+            TransactionKind::Pessimistic(_) => true,
+        }
+    }
+}
+
+impl From<TimeStamp> for TransactionKind {
+    fn from(ts: TimeStamp) -> TransactionKind {
+        match ts.try_into() {
+            Ok(ts) => TransactionKind::Pessimistic(ts),
+            Err(_) => TransactionKind::Optimistic,
+        }
+    }
+}
+
+impl From<TransactionKind> for TimeStamp {
+    fn from(tk: TransactionKind) -> TimeStamp {
+        match tk {
+            TransactionKind::Optimistic => TimeStamp::zero(),
+            TransactionKind::Pessimistic(ts) => ts.into(),
+        }
+    }
+}
+
```

Comment:
I've removed the TransactionKind -> TimeStamp conversion (I agree it is a bit weird). HOwever, I think the TimeStamp to TransactionKind conversion is useful - explicit methods are not great because then the caller has to explicitly check for zero timestamps whereas I want to make that check implicit (and have it in only one place). I could make this an inherent method with a more descriptive name if you like (although I personally think implementing From is fine).

Reply:
Personally I think a check inside the function is fine. If `TransactionKind::pessimistic(ts)` is called with `ts = 0`, let it panic, then we can easily find we make a bug. Such a `From` is error prone. If we expect a ts to be the "for update ts" of a pessimistic transaction and it is zero by mistake, we just convert it to an `Optimistic` and it's hard for us to find the mistake.

Reply:
But there are many places we don't know what kind of transaction we're in (initial decoding of the grpc request, tests, etc).